packages/components/src/date-time/date/index.tsx

```diff
@@ -100,139 +61,255 @@ function DatePickerDay( { day, events = [] }: DatePickerDayProps ) {
+								day={ day }
+								column={ index + 1 }
+								isSelected={ isSelected( day ) }
+								isFocusable={ isEqual( day, focusable ) }
+								isFocusAllowed={ isFocusAllowed }
+								isToday={ isSameDay( day, new Date() ) }
+								isInvalid={
+									isInvalidDate ? isInvalidDate( day ) : false
+								}
+								numEvents={
+									events.filter( ( event ) =>
+										isSameDay( event.date, day )
+									).length
+								}
+								onClick={ () => {
+									setSelected( [ day ] );
+									setFocusable( day );
+									onChange?.(
+										format(
+											// Don't change the selected date's time fields.
+											new Date(
+												day.getFullYear(),
+												day.getMonth(),
+												day.getDate(),
+												date.getHours(),
+												date.getMinutes(),
+												date.getSeconds(),
+												date.getMilliseconds()
+											),
+											TIMEZONELESS_FORMAT
+										)
+									);
+								} }
+								onKeyDown={ ( event ) => {
+									let nextFocusable;
+									if ( event.key === 'ArrowLeft' ) {
+										nextFocusable = subDays( day, 1 );
+									}
+									if ( event.key === 'ArrowRight' ) {
+										nextFocusable = addDays( day, 1 );
+									}
+									if ( event.key === 'ArrowUp' ) {
+										nextFocusable = subWeeks( day, 1 );
+									}
+									if ( event.key === 'ArrowDown' ) {
+										nextFocusable = addWeeks( day, 1 );
+									}
+									if ( nextFocusable ) {
+										setFocusable( nextFocusable );
+										if (
+											! isSameMonth(
+												nextFocusable,
+												viewing
+											)
+										) {
+											setViewing( nextFocusable );
+											onMonthPreviewed?.(
+												format(
+													nextFocusable,
+													TIMEZONELESS_FORMAT
+												)
+											);
+										}
+									}
+								} }
+							/>
+						);
+					} )
+				) }
+			</Calendar>
+		</div>
+	);
+}
 
-		onChange?.( newDate.set( momentTime ).format( TIMEZONELESS_FORMAT ) );
+type DayProps = {
+	day: Date;
+	column: number;
+	isSelected: boolean;
+	isFocusable: boolean;
+	isFocusAllowed: boolean;
+	isToday: boolean;
+	numEvents: number;
+	isInvalid: boolean;
+	onClick: () => void;
+	onKeyDown: ( event: KeyboardEvent ) => void;
+};
 
-		// Keep focus on the date picker.
-		keepFocusInside();
-	};
+function Day( {
+	day,
+	column,
+	isSelected,
+	isFocusable,
+	isFocusAllowed,
+	isToday,
+	isInvalid,
+	numEvents,
+	onClick,
+	onKeyDown,
+}: DayProps ) {
+	const ref = useRef< HTMLButtonElement >();
 
-	const getEventsPerDay = ( day: Moment ) => {
-		if ( ! events?.length ) {
-			return [];
+	// Focus the button when an arrow key is pressed, but only when focus was
+	// already within the calendar grid. This prevents stealing focus from an
+	// external field e.g. one of the TimePicker fields.
+	useEffect( () => {
+		if ( ref.current && isFocusable && isFocusAllowed ) {
+			ref.current.focus();
 		}
+	}, [ isFocusable ] );
```

Comment:
I guess the alternative is to set focus in the `onKeyDown` event handler directly. Is that something you considered? It'd mean there's no need for `isFocusAllowed`.

I guess you might have to use a `data` attribute or something to find the appropriate `Day` element in the DOM that way, and the event handler would probably need to be bound to the parent month.

It works nicely as it is, so no need to change it, just curious.

Should `isFocusAllowed` be in the deps?

Reply:
I'm glad you're here! This keyboard stuff took me a day to get right and the whole time I was thinking of you and your struggles with `RovingTabIndex` 😂 

> I guess the alternative is to set focus in the `onKeyDown` event handler directly. Is that something you considered? It'd mean there's no need for `isFocusAllowed`.

Yeah, I did think about it, but it makes it difficult to implement the nice behaviour where if you arrow past the beginning/end of the grid it automatically switches to the previous/next month. You have to wait until the new month renders before calling `focus()` since otherwise the element won't exist in the DOM yet. You get this for free when using this state approach since React will call `useEffect()` after rendering the new state.

> Should isFocusAllowed be in the deps?

If we did this then we'd be calling `focus()` on an element that already has focus. I suppose it doesn't actually matter... idk, I like that the intention here is explicit which is that we only want to focus the element when `isFocusable` changes to `true` 😀

Reply:
Might be worth adding a comment to the code to explain why it's missing from the deps. It's something a well meaning dev would change.

Reply:
OK, I tried to explain it all using some comments, although it feels like I just wrote the word "focus" a dozen times in a row 😅